### PR TITLE
ast/compile: also replace declared vars in with's target

### LINKED ISSUE #6979: variables in closure of every block are sometimes not recognized
## Short description 

Sometimes when using the `every` construct, variables from the outer scope are not recognized.

OPA version =  0.67.0

## Steps To Reproduce

Here is one reproduction https://github.com/bluebrown/opa-every-with-issue
See https://openpolicyagent.slack.com/archives/CBR63TK2A/p1725004042414319 for more context and examples

### PR DESCRIPTION
**Pull request by srenatus**
It might be uncommon to do this, but it's not wrong to expect this to work.

Fixes #6979.

### CODE REVIEW

**Review by anderseknert — DISMISSED** [review]
👏

**Comment by srenatus** [comment]
This isn't ready yet: we need to keep `with input as` and `with input.foo as` intact even if a local variable is shadowing `input`. I'll keep working on it if nobody beats me to a good solution.

**Comment by srenatus** [comment]
This is *RFR* now 🥳

File: ast/compile.go [hunk 1]
```diff
@@ -5466,20 +5465,38 @@ func rewriteDeclaredVarsInTerm(g *localVarGenerator, stack *localDeclaredVars, t
...
 	return errs
 }
 
+func rewriteDeclaredVarsInWithRecursive(g *localVarGenerator, stack *localDeclaredVars, w *With, errs Errors, strict bool) Errors {
+	// NOTE(sr): `with input as` and `with input.a.b.c as` are deliberately skipped here: `input` could
+	// have been shadowed by a local variable/argument but should NOT be replaced in the `with` target.
+	//
+	// We cannot drop `input` from the stack since it's conceivable to do `with input[input] as` where
+	// the second input is meant to be the local var. It's a terrible idea, but when you're shadowing
+	// `input` those might be your thing.
```
Comment (anderseknert) [inline ast/compile.go:5482]: 🤣

**Review by anderseknert — APPROVED** [review]

File: ast/compile_test.go [hunk 2]
```diff
@@ -5673,6 +5678,40 @@ func TestRewriteDeclaredVars(t *testing.T) {
 				p { __local1__ = data.test.y; data.test.q[[__local1__, __local0__]] }
 			`,
 		},
+		{
+			note: "with: rewrite target",
+			module: `
+				package test
+				p {
+					x := "foo"
+					true with input[x] as 1
```
[the remark this answers is not in the document]
Comment (srenatus) [reply ast/compile_test.go:5687]:
```
compile_test.go:6252: 1 error occurred: test.rego:4: rego_unsafe_var_error: var x is unsafe
```
Safety checks aren't touched here, yup.

**Pull request merged**